csi/controller_server.go

```diff
@@ -268,8 +268,13 @@ func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *cs
 		return nil, status.Errorf(codes.Aborted, "The volume %s is %s", req.GetVolumeId(), existVol.State)
 	}
 
+	// Check volume frontend settings
+	if existVol.DisableFrontend || existVol.Frontend != string(types.VolumeFrontendBlockDev) {
+		return nil, status.Errorf(codes.InvalidArgument, "ControllerPublishVolume: there is no block device frontend for volume %s", req.GetVolumeId())
+	}
+
 	// the volume is already attached make sure it's to the same node as this
-	if existVol.State == string(types.VolumeStateAttached) {
+	if existVol.State == string(types.VolumeStateAttached) && existVol.Controllers[0].Endpoint != "" {
```

Comment:
in the case where endpoint == ""
we are going to call attach which isn't valid in the case where volume.State == attached.
see https://github.com/longhorn/longhorn-manager/blob/master/api/model.go#L910
this will then throw an internal error, below in the else clause.